pennylane/_qubit_device.py

```diff
@@ -112,83 +112,246 @@ def execute(self, queue, observables, parameters=None):
 
             # Ensures that a combination with sample does not put
             # expvals and vars in superfluous arrays
-            if all(obs.return_type is Sample for obs in observables):
-                return self._asarray(results)
-            if any(obs.return_type is Sample for obs in observables):
+            sample_return_types = (obs.return_type is Sample for obs in observables)
+            if any(sample_return_types) and not all(sample_return_types):
                 return self._asarray(results, dtype="object")
 
             return self._asarray(results)
 
-    def rotate_basis(self, observable):
+    def statistics(self, observables):
+        """Process measurement results from circuit execution and return statistics.
+
+        This includes returning expectation values, variance, samples and probabilities.
+
+        Args:
+            observables (Union[:class:`Observable`, List[:class:`Observable`]]): the number of basis states to sample from
+
+        Raises:
+            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
+
+        Returns:
+            Union[float, List[float]]: the corresponding statistics
+        """
+        results = []
+
+        for obs in observables:
+            # Pass instances directly
+            if obs.return_type is Expectation:
+                results.append(self.expval(obs))
+
+            elif obs.return_type is Variance:
+                results.append(self.var(obs))
+
+            elif obs.return_type is Sample:
+                results.append(np.array(self.sample(obs)))
+
+            elif obs.return_type is Probability:
+                results.append(self.probability(wires=obs.wires))
+
+            elif obs.return_type is not None:
+                raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
+
+        return results
+
+    def rotate_basis(self, obs_queue):
         """Rotates the specified wires such that they
         are in the eigenbasis of the provided observable.
 
         Args:
-            observable (Observable): the observable that is to be
-                measured
+            observables (List[:class:`Observable`]): the observable queue
+        """
+        self._prob = self.probability()
+        self._memory = False
+
+        wires = []
+
+        for observable in obs_queue:
+            if hasattr(observable, "return_type") and observable.return_type == Sample:
+                self._memory = True  # make sure to return samples
+
+            for diag_gate in observable.diagonalizing_gates():
+                self.apply(diag_gate)
+
+            for wire in observable.wires:
+                if isinstance(wire, int):
+                    wires.append(wire)
+                else:
+                    wires.extend(wire)
+
+        self._wires_used = wires
```

Comment:
What is the point of `_wires_used`?

As `rotate_basis` is also called inside `expval` for example, this value will be reset continually. If other methods then depend on it this might cause unexpected behaviour.

Reply:
> What is the point of _wires_used?

I'll let @antalszava answer this one!

> As rotate_basis is also called inside expval for example

Note that calling `rotate_basis` inside `expval` is a quick 'hack' to make the existing default.qubit test suite pass (as it doesn't have a concept of rotation, and simply calls `dev.expval`). The plan is to remove this call inside `expval`, and refactor the tests to take this into account.

Reply:
The point of ``_wires_used`` is to keep track of the wires where an observable is applied. Those wires, that are not in ``_wires_used`` (if any) will have the ``Identity`` applied implicitly and probabilities will need to be summed over such wires once marginal probabilities are returned.

It is good feedback to see that it is not entirely clear how this is used, will try to rename it.

Currently   ``rotate_basis`` is only called inside of e.g. ``expval`` if the following statement evaluates as ``True``:
```python
if self._rotated_prob is None:
```
This logic will be refactored at a later stage, currently it is only included such that tests can be passed easily (it would require a major test refactor).

Reply:
Okay, I see!

Reply:
And from my point of view, `_wires_used` is a pretty good name, no need to change that. I was just wondering if the `_wires_used` attribute is still sane if it is modified in expval